src/test/java/api/loans/ScheduledDueDateNoticesProcessingTests.java

```diff
@@ -0,0 +1,385 @@
+      .toArray(Matcher[]::new);
+
+    List<JsonObject> sentNotices = patronNoticesClient.getAll();
+    assertThat(sentNotices, hasSize(expectedTemplateIds.length));
+    assertThat(sentNotices, hasItems(matchers));
+  }
+
+  @Test
+  public void beforeNoticeShouldBeSentAndItsNextRunTimeShouldBeUpdated()
+    throws MalformedURLException,
+    InterruptedException,
+    TimeoutException,
+    ExecutionException {
+
+    DateTime beforeDueDateTime = dueDate.minus(beforePeriod.timePeriod()).plusSeconds(1);
+    scheduledNoticeProcessingClient.runNoticesProcessing(beforeDueDateTime);
+
+    checkSentNotices(beforeTemplateId);
+
+    DateTime expectedNewRunTimeForBeforeNotice = dueDate
+      .minus(beforePeriod.timePeriod())
+      .plus(beforeRecurringPeriod.timePeriod());
+
+    checkScheduledNotices(
+      expectedNewRunTimeForBeforeNotice,
+      dueDate,
+      dueDate.plus(afterPeriod.timePeriod()));
+  }
+
+  @Test
+  public void beforeNoticeShouldBeSendAndDeletedWhenItsNextRunTimeIsAfterDueDate()
+    throws MalformedURLException,
+    InterruptedException,
+    TimeoutException,
+    ExecutionException {
+
+    DateTime justBeforeDueDateTime = dueDate.minusSeconds(1);
+    scheduledNoticeProcessingClient.runNoticesProcessing(justBeforeDueDateTime);
+
+    checkSentNotices(beforeTemplateId);
+
+    checkScheduledNotices(
+      null,
+      dueDate,
+      dueDate.plus(afterPeriod.timePeriod()));
+  }
+
+  @Test
+  public void beforeAndUponAtNoticesShouldBeSentWhenProcessingJustAfterDueDate()
+    throws MalformedURLException,
+    InterruptedException,
+    TimeoutException,
+    ExecutionException {
+
+    DateTime justAfterDueDateTime = dueDate.plusSeconds(1);
+    scheduledNoticeProcessingClient.runNoticesProcessing(justAfterDueDateTime);
+
+    checkSentNotices(beforeTemplateId, uponAtTemplateId);
+
+    checkScheduledNotices(
+      null,
+      null,
+      dueDate.plus(afterPeriod.timePeriod()));
+  }
+
+  @Test
+  public void afterRecurringNoticeShouldBeSentSeveralTimesBeforeLoanIsClosed()
+    throws MalformedURLException,
+    InterruptedException,
+    TimeoutException,
+    ExecutionException {
+
+    DateTime justAfterDueDateTime = dueDate.plusSeconds(1);
+    scheduledNoticeProcessingClient.runNoticesProcessing(justAfterDueDateTime);
+    //Clear all sent notices before actual test
+    patronNoticesClient.deleteAll();
+
+    DateTime afterNoticeRunTime = dueDate.plus(afterPeriod.timePeriod()).plusSeconds(1);
+    scheduledNoticeProcessingClient.runNoticesProcessing(afterNoticeRunTime);
+
+    DateTime afterNoticeExpectedRunTime = dueDate
+      .plus(afterPeriod.timePeriod())
+      .plus(afterRecurringPeriod.timePeriod());
+
+    checkScheduledNotices(
+      null,
+      null,
+      afterNoticeExpectedRunTime);
+
+    //Run again to send recurring notice
+    scheduledNoticeProcessingClient.runNoticesProcessing(
+      afterNoticeExpectedRunTime.plusSeconds(1));
+
+    checkSentNotices(afterTemplateId, afterTemplateId);
```

Comment:
Are there any parameters or tokens that should be different between the two notices, e.g. the date of the notice?

Reply:
The rest parameters appear to be the same.

Reply:
What do you mean by the rest parameters?

I guess what I'm trying to understand is, if a reoccurring after notice is set up, what should be different between the two notices, from the receivers perspective?

Reply:
From receivers perspective the only difference is time of arrival. But there is no difference in representation itself.